### ISSUE TITLE
`QReg`s not permuted by `pecos.slr.Permute`

### ISSUE THREAD

**Issue by perlinm** [PECOS-packages/PECOS#95]
It looks like `pecos.slr.Permute` correctly permutes individual qubits, but does not relabel qubit registers.  What I mean is best understood by an example:
```python
#!/usr/bin/env python3
import pecos.qeclib.qubit
import pecos.slr

qreg_a = pecos.slr.QReg("qreg_a", 1)
qreg_b = pecos.slr.QReg("qreg_b", 1)
program = pecos.slr.Main(
    pecos.slr.Permute(qreg_a, qreg_b),
    pecos.qeclib.qubit.X(qreg_a),
    pecos.qeclib.qubit.X(qreg_a[0]),
)

print(program.qasm())
```
Based on my understanding of how `Permute` is supposed to work, I expect this script to print X gates on `qreg_b` and `qreg_b[0]`.  However, on the current `PECOS:development` branch (commit `eca0ce7af8057a7fbd6e512d499081e8bb9bffad`), the above script instead prints
```sh
OPENQASM 2.0;
include "hqslib1.inc";
// Generated using: PECOS version 0.6.0.dev6
// Permuting: qreg_b[0] -> qreg_a[0]
x qreg_a;
x qreg_b[0];
```
so the qubit is being "relabeled" properly, but the register is not.  Regardless of my prior expectations, this behavior does not seem to be self-consistent.  Any use of `Permute` therefore messes up all following QASM in a `pecos.slr.Main` program.

In particular, this issue currently prevents us from running logical circuits with (1) the teleportation variant of a T gate ([`Steane.t_tel`](https://github.com/PECOS-packages/PECOS/blob/eca0ce7af8057a7fbd6e512d499081e8bb9bffad/python/pecos/qeclib/steane/steane_class.py#L270-L285)), and (2) Steane-type QEC with one-bit teleportation.

Not sure if this helps, but the issue **might** be in how qubits/registers/etc are permuted in [`gen_qasm.process_permute`](https://github.com/PECOS-packages/PECOS/blob/eca0ce7af8057a7fbd6e512d499081e8bb9bffad/python/pecos/slr/gen_codes/gen_qasm.py#L348).  The circular references between `QReg`s and `Qubit`s make it a bit difficult for me to reason about how to solve the issue, though...

FWIW, I think using `Permute` would be unnecessary if hardware could "natively" recognize a virtual SWAP instruction in QASM, as in something like `swap qreg_0[0] qreg_1[0];`.  That would be both a nice feature to have and a more elegant solution than permuting register names in QASM.

**Comment by perlinm** [comment]
More weirdness with `pecos.slr.Permute`: if I print
```python
print(qreg_a)
print(qreg_a.elems[0].reg)
```
in the script above, I get
```sh
qreg_a
qreg_b
```
so register `qreg_a` has a qubit that thinks it belongs to `qreg_b`.  This is expected in light of how `gen_qasm.process_permute` [works](https://github.com/PECOS-packages/PECOS/blob/eca0ce7af8057a7fbd6e512d499081e8bb9bffad/python/pecos/slr/gen_codes/gen_qasm.py#L355-L357), but it does not seem to be "consistent" behavior.  Registers and qubits are not "on the same page" about ownership.

It seems to me like permuting individual qubits cannot (at least cleanly) be made consistent with the fact that QASM allows for broadcasting of the form `x qreg_a;`.  If I have two multi-qubit registers, and I want to permute the first qubits of these registers (`Permute(qreg_a[0], qreg_b[0])`), that simply cannot be achieved by relabeling registers in a QASM file (which I think is essentially how PECOS tries to `Permute`).

Possible solutions off the top of my mind:
1. Only allow permuting registers (of equal size) in PECOS.  This is somewhat heavy-handed, but prevents the sorts of issues discussed above.  Permuting two registers would simply make them swap names (`QReg.sym`).
2. "Unroll" all broadcasting when generating QASM, i.e., make commands like `pecos.qeclib.qubit.X(qreg_a)` print an `x` gate on each individual qubit of `qreg_a`.  This seems... inelegant but doable I suppose?
3. Add hardware-native (virtual) SWAP gates, and replace all uses of `pecos.slr.Permute` with SWAP gates 🙃

**Comment by ciaranra** [comment]
Weird, I could of swore I check permuting whole registers a week or two ago. Maybe this was an oversight that wasn't tested with the new Qasm code generation update.

Also, yeah. Need to unroll things like `X q;` for this reason. I remember I meant to do this for this reason but looks like that slipped.

Permuting different sized registers is already not allowed.

It appears the qubit/bit Permuting is still working, which could be used as a work around for now:

```python
from pecos.qeclib import qubit as Q
from pecos.slr import CReg, QReg, Main, Permute

prog = Main(
    a := QReg("a", 1),
    b := QReg("b", 1),

    Permute(
        [a[0], b[0]], 
        [b[0], a[0]]
    ),
    Q.X(a[0]),
    Q.X(b[0]),
    
)

print(prog.qasm())
```

produces:
```
OPENQASM 2.0;
include "hqslib1.inc";
// Generated using: PECOS version 0.6.0.dev6
qreg a[1];
qreg b[1];
// Permuting: b[0] -> a[0], a[0] -> b[0]
x b[0];
x a[0];
```

**Comment by perlinm** [comment]
Minor clarification: when I said "only allow permuting registers (of equal size) in PECOS", I was proposing to *forbid* permuting individual qubits.  In any case, I opened [PR#96](https://github.com/PECOS-packages/PECOS/pull/96) with my second proposed fix (unrolling broadcast operations when producing QASM), and I still ultimately think the third fix (supporting virtual SWAP gates) is best in the long term.

**Comment by perlinm** [comment]
fixed by #96
